tla/issue.md

```diff
@@ -0,0 +1,174 @@
+3. The backup (at view 0) replica 1 receives the `PrepareRequest` of view 0 and broadcasts its `PrepareResponse`.
+4. The backup (at view 0) replica 1 decides to change its view (possible on timeout) and sends the `ChangeView` message.
+5. The backup (at view 0) replica 2 receives the `PrepareRequest` of view 0 and broadcasts its `PrepareResponse`.
+6. The backup (at view 0) replica 2 collects `M` prepare messages (from itself and replicas 0, 1) and broadcasts the `Commit` message for view 0.
+7. The backup (at view 0) replica 3 decides to change its view (possible on timeout) and sends the `ChangeView` message.
+
+Here's the TLC error traces attached:
+ * [TLC error trace for the first configuration (primary is "dying")](./base_deadlock2_dl.txt)
+ * [TLC error trace for the second configuration (non-primary is "dying")](./base_deadlock3_dl.txt)
+
+</details>
+
+After the described sequence of steps we end up in the following situation:
+ 
+| Replica | View | State                                                          |
+|---------|------|----------------------------------------------------------------|
+| 0       | 0    | Dead (but `PrepareRequest` sent for view 0).                   |
+| 1       | 0    | `ChangeView` sent, in the process of changing view from 0 to 1 |
+| 2       | 0    | `Commit` sent, waiting for the rest nodes to commit at view 0  |
+| 3       | 0    | `ChangeView` sent, in the process of changing view from 0 to 1 |
+
+So we have the replica 0 permanently dead at the view 0 without possibility to affect
+the consensus process. Replica 2 has its `Commit` sent and unsuccessfully waiting for
+other replicas to enter the commit stage as far. Finally, replicas 1 and 3 have entered
+the view changing stage and not being able either to commit (as there's only `F` nodes
+that have been committed at the view 1) or to change view (as the replica 2 can't send
+its `ChangeView` from the commit stage).
+
+It should be noted that dBFT 2.0 algorithm is expected to guarantee the block acceptance
+with at least `M` non-faulty ("alive" in this case) nodes, which isn't true in the
+described situation.
+
+#### 3. Running the TLC model checker with "faulty" nodes
+
+Both models allow to specify the set of malicious nodes indexes via `RMFault` model
+constant. "Faulty" nodes are allowed to send *any* valid message at *any* step of
+the behaviour. At the same time, weak fairness is required from the next-state
+action predicate for both models. It means if it's possible for the model to take
+any non-stuttering step, this step must eventually be taken. Thus, running the
+basic dBFT 2.0 model with single faulty and three non-faulty nodes doesn't reveal
+any model deadlock: the malicious node keep sending messages if possible to escape
+from the liveness lock. Moreover, the presence of faulty nodes significantly
+increases the states graph size, so that it takes quite a long time to evaluate
+the whole set of possible model behaviours.
```

Comment:
Haven't we checked this completely?

Reply:
Yes, we have for both scenarious when primary and non-primary is faulty. So I probably have to remove the last sentence.

Reply:
Actually, I still have doubts about it and not sure whether it works correctly. I really expected this faulty model to end up in a deadlock.